test/watcher_test.js

```diff
@@ -139,6 +162,65 @@ describe('Watcher', function() {
       expect(builderBuild).to.have.been.called;
     });
 
+    it('on change, rebuild is invoked and cancel is invoked from the build', function() {
+      const pluginsCalled = {};
+
+      const waitForPlugin1 = defer();
+      const waitForPlugin2 = defer();
+
+      class Plugin1 extends Plugin {
+        constructor(inputNodes, options) {
+          super(inputNodes || [], options);
+        }
+
+        build() {
+          pluginsCalled['plugin1'] = pluginsCalled['plugin1'] ? pluginsCalled['plugin1'] + 1 : 1;
+
+          return waitForPlugin1.promise;
+        }
+      }
+
+      class Plugin2 extends Plugin {
+        constructor(inputNodes, options) {
+          super(inputNodes || [], options);
+        }
+
+        build() {
+          pluginsCalled['plugin2'] = pluginsCalled['plugin2'] ? pluginsCalled['plugin2'] + 1 : 1;
+
+          return waitForPlugin2.promise;
+        }
+      }
+
+      const plugin1 = new Plugin1();
+      const plugin2 = new Plugin2();
+      const builder = new Builder(Merge([plugin1, plugin2]));
+      const watcher = new Watcher(builder, [watchedNodeBasic], { watcherAdapter: adapter });
+
+      watcher._ready = true;
+      watcher.start();
+
+      return sleep()
```

Comment:
Why are we `sleep`ing? What are we waiting for here?

Reply:
I just wanted to start a promise chain. I can just do Promise.resolve().then()

Reply:
This new approach looks good to me, I think a small comment could likely address @rwjblue's concern.

It would seem we could rewrite the rest of this function as a chain of `awaits`

```js
await Promise.resolve() // wait for 1 turn of the micro-task queue
{
   const build = watcher._change('change', 'foo.js', 'root');

          waitForPlugin1.resolve();
          waitForPlugin2.resolve();

  await build;
} 

// ...
```

Reply:
I make this waitForTheWatcher to be ready. Basically it needs to boot up before we can do anything to it. We could run a full build for the watcher but then I would have to wire up the promises to only return on the second pass. This seems cleaner and easier to follow. Thoughts?